Build make windows

jheaff1 opened the pull request:
This pull request fixes the issue of GNU Make failing to build on Windows

UebelAndre commented:
Just FYI, I believe we don't have `make` building on windows because we don't have a target running in CI that uses `make`. I'd recommend deleting the [`- "-//make_simple/..."`](https://github.com/bazelbuild/rules_foreign_cc/blob/bb2f0ab0aa7bf1cbcd4836dbfca353f06fd38c41/.bazelci/config.yaml#L110) line from the job definition to enable a target that would start building it:
https://github.com/bazelbuild/rules_foreign_cc/blob/bb2f0ab0aa7bf1cbcd4836dbfca353f06fd38c41/.bazelci/config.yaml#L95-L111

jheaff1 commented:
@UebelAndre  I have re-enabled the `make_simple` test but it seems to be failing. Any ideas?

UebelAndre commented:
Ugh, these are so annoying:
```
liba.a(liba.o) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in test_libb.obj
```

I think you need to pass something like:
```
CFLAGS=-MT CXXFLAGS=-MT
```

I've only ever dealt with this in CMake so not quite sure how best to wire this up...

UebelAndre commented:
Another potential solution may be:
https://github.com/bazelbuild/rules_foreign_cc/blob/bb2f0ab0aa7bf1cbcd4836dbfca353f06fd38c41/examples/cmake_with_data/BUILD.bazel#L7-L12

But you can go either way. Statically liking might be better.

jheaff1 commented:
@UebelAndre  It's passing now, thanks!

UebelAndre commented:
I wonder how many more tests you can re-enable now that make is buildable 😄

jsharpe commented on foreign_cc/built_tools/private/built_tools_framework.bzl:
```diff
@@ -1,12 +1,17 @@
 # Common attributes for all built_tool rules
 FOREIGN_CC_BUILT_TOOLS_ATTRS = {
+    "env": attr.string_dict(
+        doc = "Environemnt variables to set during the build.",
```
Nit: Typo in environment

UebelAndre replied:
```suggestion
        doc = "Environment variables to set during the build.",
```

jsharpe commented on foreign_cc/built_tools/private/built_tools_framework.bzl:
```diff
@@ -44,12 +54,18 @@ def built_tool_rule_impl(ctx, script_lines, out_dir, mnemonic):
 
-    script = [
+    cc_toolchain = find_cpp_toolchain(ctx)
+    linker_path_str = "\"" + paths.dirname(cc_toolchain.ld_executable).replace("C:", "/c") + "\""
```
What if the build is executed from a drive other than C?

UebelAndre replied:
Would it be possible to use a relative path? If the env prelude from the common framework is being used, you could probably prefix with `$EXT_BUILD_ROOT` or something. But I do think we can't assume `C` drive here.

jsharpe reviewed (commented):
LGTM just a couple of small comments

jheaff1 commented:
> I wonder how many more tests can be re-enable now that make is buildable 😄
> 
> @jheaff1 Would you mind doing a quick check?

I added some temp commits to check which examples (including the third party examples) the CI could build. I have modified the most recent commit to enable the tests that passed (namely, `cmake_working_dir` and the `curl_build_test`)

UebelAndre commented on foreign_cc/built_tools/private/built_tools_framework.bzl:
```diff
@@ -44,12 +55,25 @@ def built_tool_rule_impl(ctx, script_lines, out_dir, mnemonic):
+    path_prepend_cmd = ""
+    if "win" in os_name(ctx):
+        # Prepend PATH environment variable with the path to the toolchain linker, which prevents MSYS using its linker (/usr/bin/link.exe) rather than the MSVC linker (both are named "link.exe")
+        linker_path = paths.dirname(cc_toolchain.ld_executable)
```
Will it always be the case that this path starts with a drive? I feel like it'd be worth checking for that case but then otherwise doing nothing. Your thoughts?
```suggestion
        linker_path = paths.dirname(cc_toolchain.ld_executable)
        if linker_path[0].is_alpha() and linker_path[1] == ":":
            linker_path = linker_path.replace(linker_path[0:2], "/" + linker_path[0].lower())
```

jheaff1 replied:
Sure, done :)

UebelAndre reviewed (approved):
This is a huge improvement! This (to me) just unblocked support for an entire platform. Thanks for submitting this!!

jheaff1 commented:
No problem at all, glad I could help :)
